src/System.Memory/src/System/Buffers/Text/Utf8Formatter/Utf8Formatter.Integer.Signed.Default.cs

```diff
@@ -13,50 +11,45 @@ public static partial class Utf8Formatter
         {
-            ref byte utf8Bytes = ref buffer.DangerousGetPinnableReference();
             int idx = 0;
 
             if (value < 0)
             {
                 if (buffer.Length < 2) goto FalseExit;  // Buffer of length 1 won't have space for the digits after the minus sign
-                Unsafe.Add(ref utf8Bytes, idx++) = Utf8Constants.Minus;
+                buffer[idx++] = Utf8Constants.Minus;
 
                 // Abs(long.MinValue) == long.MaxValue + 1, so we need to handle this specially.
                 if (value == long.MinValue)
                 {
-                    if (buffer.Length < 20) goto FalseExit; // WriteDigits does not do bounds checks
-                    Unsafe.Add(ref utf8Bytes, 1) = (byte)'9';
-                    FormattingHelpers.WriteDigits(223372036854775808L, 18, ref utf8Bytes, 2);
-                    bytesWritten = 20;
+                    // digitCount + idx = 19 + 1 = 20
+                    if (buffer.Length < 19 + idx) goto FalseExit;  // WriteDigits does not do bounds checks
+                    buffer[idx] = (byte)'9';
+                    // Already wrote '9', 19 - 1 = 18 digits left
+                    FormattingHelpers.WriteDigits(223372036854775808L, 18, ref buffer.DangerousGetPinnableReference(), idx + 1);
+                    bytesWritten = 19 + idx;
                     return true;
                 }
 
                 value = -value;
             }
 
-            long left = value;
-            for (int i = idx; i < buffer.Length; i++)
+            if (value < 10)
             {
-                left = FormattingHelpers.DivMod(left, 10, out long num);
-                Unsafe.Add(ref utf8Bytes, i) = (byte)('0' + num);
-                if (left == 0)
-                {
-                    i++;
-                    // Reverse the bytes
-                    for (int j = 0; j < ((i - idx) >> 1); j++)
-                    {
-                        byte temp = Unsafe.Add(ref utf8Bytes, j + idx);
-                        Unsafe.Add(ref utf8Bytes, j + idx) = Unsafe.Add(ref utf8Bytes, i - j - 1);
-                        Unsafe.Add(ref utf8Bytes, i - j - 1) = temp;
-                    }
-                    bytesWritten = i;
-                    return true;
-                }
+                if (buffer.Length == 0) goto FalseExit;
+                buffer[idx] = (byte)('0' + value);
+                bytesWritten = 1 + idx;
+                return true;
             }
 
+            int digitCount = FormattingHelpers.CountDigits((ulong)value);
+            if (buffer.Length < digitCount + idx) goto FalseExit;  // WriteDigits does not do bounds checks
+            FormattingHelpers.WriteDigits(value, digitCount, ref buffer.DangerousGetPinnableReference(), idx);
+            bytesWritten = digitCount + idx;
+            return true;
+
+        FalseExit:
             // Buffer too small, clean up what has been written
             buffer.Clear();
```

Comment:
You just need to clear exactly one byte - no need to call Clear for it. (It would be even nicer to avoid the need for clearing completely.)

Reply:
> It would be even nicer to avoid the need for clearing completely.

Why is there a need?  Seems like we could just delete this line.

Reply:
I have tried to write fast and safe TryFormatInt64Default - here is what I got (untested):

```C#
private static bool TryFormatInt64Default(long value, Span<byte> buffer, out int bytesWritten)
{
    if ((ulong)value < 10)
    {
        if (buffer.Length != 0) goto FalseExit;
        buffer[0] = (byte)('0' + value);
        bytesWritten = 1;
        return true;
    }

    int digits = CountDigits((ulong)((value < 0) ? -value : value));
    if (value < 0)
    {
        value = -value;
        if (digits >= buffer.Length) goto FalseExit;
        bytesWritten = digits+1;
        buffer[0] = (byte)'-';
        buffer = buffer.Slice(1);
    }
    else
    {
        if (digits > buffer.Length) goto FalseExit;
        bytesWritten = digits;
    }
    ulong left = (ulong)value;
    for (int i = digits - 1; i >= 1; i--)
    {
        left = DivMod(left, 10, out ulong num);
        buffer[i] = (byte)('0' + num);
    }
    buffer[0] = (byte)('0' + left);
    return true;

FalseExit:
    bytesWritten = 0;
    return false;
}

// Not sure whether it is worth it to mark CountDigits with AggressiveInlining. Need to measure.
[MethodImpl(MethodImplOptions.AggressiveInlining)]
private static int CountDigits(ulong value)
{
...
}
```

Reply:
This works really well in most cases, but causes a 10% regression for the variable length test:
![image](https://user-images.githubusercontent.com/6527137/33347442-b2e64cb0-d447-11e7-87b4-962c8ec5eccb.png)


```C#
private static readonly int[] s_Int32TextArray = new int[20]
{
        214748364,
        2,
        21474836,
        -21474,
        21474,
        -21,
        -2,
        214,
        -21474836,
        -214748364,
        2147,
        -2147,
        -214748,
        -2147483,
        214748,
        -2147483648,
        2147483647,
        21,
        2147483,
        -214
};

private static void Utf8IntFormatter_VariableLength()
{
    int textLength = s_Int32TextArray.Length;
    Span<byte> buffer = new byte[100];
    foreach (var iteration in Benchmark.Iterations)
    {
        using (iteration.StartMeasurement())
        {
            for (int i = 0; i < Benchmark.InnerIterationCount; i++)
            {
                bool retVal = Utf8Formatter.TryFormat(s_Int32TextArray[i % textLength], buffer, out int bytesWritten);
                DoNotIgnore(ref buffer, bytesWritten, retVal);
            }
        }
    }
}
```

Reply:
Also, if we change to unsafe, we get a 5-10% performance gain:
```C#
    ref byte utf8Bytes = ref buffer.DangerousGetPinnableReference();
    ulong left = (ulong)value;
    for (int i = digits - 1; i >= 1; i--)
    {
        left = DivMod(left, 10, out ulong num);
        Unsafe.Add(ref utf8Bytes, i) = (byte)('0' + num);
    }
    utf8Bytes = (byte)('0' + left);
```

![image](https://user-images.githubusercontent.com/6527137/33347616-4cfed84e-d448-11e7-9997-1bab09c0e10a.png)

> // Not sure whether it is worth it to mark CountDigits with AggressiveInlining. Need to measure.

Inlining CountDigits helps:
![image](https://user-images.githubusercontent.com/6527137/33347680-8f59bd26-d448-11e7-88a6-f0dab55eae00.png)

Reply:
> Also, if we change to unsafe, we get a 5-10% performance gain:

Yes, you will be always able to get some extra gains from unsafe code.

My point was that if you spend some effort optimizing the safe implementation, you can beat the gains you get from unsafe code. And with safe code, you can save time on security sign-off activities for unsafe code and instead spend time on optimizing the safe implementation even further.

For example, the performance of these formatters is going to be really bad (like multiple times bad) on 32-bit platforms because of they always run in 64-bit. It would be nice to fix that.

Reply:
> 10% regression for the variable length test

That's weird. Looks like a micro-architecture hole.